src/Expander.hs

```diff
@@ -882,7 +1001,17 @@ runTask (tid, localData, task) = withLocal localData $ do
                     then return tid
                     else newTaskID
           laterMacro tid' b v x dest newDeps mdest stx
-        [] ->
+        [] -> do
+          inEarlierPhase $
+            forkCompleteTypeCheck mdest $
+              Ty (TFun (Ty TSyntax)
+                       (Ty (TMacro (Ty TSyntax))))
+          forkAwaitingMacroType b v x mdest dest stx
+    AwaitingMacroType dest (TaskAwaitMacroType b v x mdest stx) ->
+      isExprChecked mdest >>=
+      \case
+        False -> stillStuck tid task
```

Comment:
Just thinking out loud, but I feel like the optimal API for tasks would be two monadic functions: one in a `Reader` to check readiness/preconditions, and another in `State` to actually do the thing. That way, the expander could check what tasks are ready without potentially mutating state.

Reply:
One downside to that approach is that sometimes the check for "can I run" makes a little progress. For instance, `AwaitingMacro` has a list of roots that need to be fully filled in, but when it checks whether it can run, it recalculates the fringe of the trees to avoid retraversals. I don't know that we can do this with just a `Reader`-style effect.